readme.md
# LaravelSqlines

[![Status Release](https://img.shields.io/badge/Status-Beta-red)]() 
[![Latest Version on Packagist][ico-version]][link-packagist]
[![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://opensource.org/licenses/MIT) 


Simple laravel package to manage the sync of databases over [Sqlines](http://www.sqlines.com/).


## Installation

Via Composer

``` bash
$ composer require brstuff/laravelsqlines
```

Copy config and sqlines to storage

```bash
php artisan vendor:publish --tag=laravelsqlines.config
php artisan vendor:publish --tag=laravelsqlines.app
```


## Usage

Simple example. 

```php
$log = \LaravelSqlines::
    sourceConnection("oracle", "database_name_in_config")
    ->targetConnection("mysql", "database_name_in_config")
    ->tables(["table1", "tables2"])
    ->tmap("table1, new_table_1")
    ->tmap("table2, new_table_2")
    ->sync();
```

## License

MIT License (MIT). Please see the [license file](license.md) for more information.

[ico-version]: https://img.shields.io/packagist/v/brstuff/laravelsqlines.svg?style=flat-square
[ico-travis]: https://img.shields.io/travis/brstuff/laravelsqlines/master.svg?style=flat-square
[ico-styleci]: https://styleci.io/repos/12345678/shield

[link-packagist]: https://packagist.org/packages/brstuff/laravelsqlines
[link-downloads]: https://packagist.org/packages/brstuff/laravelsqlines
[link-travis]: https://travis-ci.org/brstuff/laravelsqlines
[link-styleci]: https://styleci.io/repos/12345678
[link-author]: https://github.com/brstuff



changelog.md
# Changelog

All notable changes to `LaravelSqlines` will be documented in this file.

## Version 1.0

### Added
- Everything


config/laravelsqlines.php
<?php

return [ 
    
    
    // Paths of libraries needed to run sqlines
    // ex: /usr/lib64/mysql/:/usr/lib/oracle/10.2.0.5/client64/lib/
    'export_paths' => '/usr/lib64/mysql/:/usr/lib/oracle/10.2.0.5/client64/lib/',
    
    
    
    // SQLines Data options
    'data-options' => [
        // DDL Options
        //////////////
        
        // Set yes to to migrate table definitions, or no to not migrate
        // DDL statements executed depends on -topt option (CREATE; DROP and CREATE TABLE, or TRUNCATE TABLE)
        'ddl_tables' => '',
        
        // Set yes to to migrate constraints (this is default), or no to not migrate
        'constraints' => '',
        
        // Set yes to to migrate indexes (this is default), or no to not migrate
        'indexes' => '',
        
        // Data Options
        //////////////-
        
        // Set yes to to migrate data (this is default), or no to not migrate
        'data' => '',
        
        // The maximum number of rows in read/write batch, by default it is defined by the buffer size divided by the table row size
        'batch_max_rows' => 10000,
        
        // Work with LOBs as VARCHAR columns to increase performance (use if LOB columns contain short values less than a few MB i.e.), default is no
        'fetch_lob_as_varchar' => '',
        
        // Fixed size buffer in bytes to read LOB values by binding not by reading part by part (can cause truncation error if the buffer is less than
        // the maximum LOB value, default is to read LOBs by separate calls
        'lob_bind_buffer' => 1000000,
        
        // When converting from ASCII or UTF16/UCS-2 character sets in the source database to UTF8 i.e. in the target database depending on
        // the actual data you may need greater storage size. And vice versa converting in opposite direction you may require smaller storage size.
        // This parameter specifies the length change ratio. If the source length is 100, and ratio is 1.1 then the target length will be 110
        'char_length_ratio' => '',
        
        // MariaDB options
        //////////////////
        
        // Path to MariaDB library including the file name. For example, for MariaDB Connector C on Windows:
        //   -mariadb_lib=C:\Program Files\MariaDB\MariaDB Connector C 64-bit\lib\libmariadb.dll
        // By default, on Windows the tool tries to load libmariadb.dll library from PATH; if not found the tool tries to use MySQL connector libmysql.dll
        'mariadb_lib' => '',
        
        // Disable or enable binary logging for the connection (the client must have the SUPER privilege for this operation). By default, MariaDB default is used.
        'mariadb_set_sql_log_bin' => 0,
        
        // Run SET UNIQUE_CHECKS=value at the beginning of each session, not executed if no value set (MariaDB as target)
        'mariadb_set_unique_checks' => 0,
        
        // Set global max_allowed_packet option to the specified value (use only values in multiples of 1024, for example, 1073741824 for 1GB)
        // Use this option when you receive 'MySQL has gone away' error during the data load
        'mariadb_max_allowed_packet' => '',
        
        // MySQL options
        ////////////////
        
        // Set the character set for the connection (utf8mb4 i.e.)
        'mysql_set_character_set' => 'utf8',
        
        // Set the collation for the connection (ex utf8mb4_bin)
        'mysql_set_collation' => 'utf8_bin',
        
        // Run SET FOREIGN_KEY_CHECKS=value at the beginning of each session, not executed if no value set (MySQL as target)
        'mysql_set_foreign_key_checks' => 0,
        
        // Collate used for data validation. Use _bin or _cs collates to order values ABCabc instead of AaBbCc
        'mysql_validation_collate' => 'latin1_bin',
        
        // SQL Server options
        ////////////////////'
        
        // Codepage for input data (BCP -C option)
        'bcp_codepage' => '',
        
        // Oracle options
        ////////////////'
        
        // Path to Oracle OCI library including the file name. For example, for Oracle on Windows:
        //   -oci_lib=C:\oraclexe\app\oracle\product\11.2.0\server\bin\oci.dll
        // By default, on Windows the tool tries to load oci.dll library from PATH; if not found the tool tries to search Windows registry to find Oracle client installations
        // See also http://www.sqlines.com/sqldata_oracle_connection
        'oci_lib' => '',
        
        // NLS_LANG setting to use for Oracle connection
        'oracle_nls_lang' => 'AMERICAN_AMERICA.WE8ISO8859P1',
        
        // PostgreSQL options
        ////////////////////'
        
        // Set client encoding (you can also use PGCLIENTENCODING environment variable)
        'pg_client_encoding' => 'latin1',
        
        // Sybase ASE options
        ////////////////////'
        
        // Codepage
        'sybase_codepage' => 'cp850',
        
        // Set to yes to use encrypted password handshakes with the server
        'sybase_encrypted_password' => '',
        
        // Sybase ASA (Sybase SQL Anywhere) options
        //////////////////////////////////////////'
        
        // Set to yes to extract all character data as 2-byte Unicode (UTF-16/UCS-2)
        'sybase_asa_char_as_wchar' => '',
        
        // Informix options
        //////////////////'
        
        // Set CLIENT_LOCALE for connection (Note that Setnet32 and environment variable has no effect on Informix ODBC driver)
        'informix_client_locale' => '',
        
        // Validation Options
        ////////////////////'
        
        // Maximum number of found not equal rows per table after which the validation stops for this table, by default no limit set
        'validation_not_equal_max_rows' =>'',
        
        // Number of digits in the fractional part (milliseconds, microseconds i.e.) of datetime/timestamp values to validate. By default, all digits are compared
        'validation_datetime_fraction' => '',
        
        // Misc Options
        //////////////'
        
        // Set to yes to generate trace file sqldata.trc with debug information, default is no
        'trace' => '',
        
        // Set to yes to create dump files containing data for tables
        'trace_data' => '',
        
        // Set to yes to create dump files containing differences in data found during data validation
        'trace_diff_data' => '',
    ]
];

src/LaravelSqlinesServiceProvider.php
<?php

namespace brstuff\LaravelSqlines;

use Illuminate\Support\ServiceProvider;

class LaravelSqlinesServiceProvider extends ServiceProvider
{
    /**
     * Perform post-registration booting of services.
     *
     * @return void
     */
    public function boot()
    {
        // $this->loadTranslationsFrom(__DIR__.'/../resources/lang', 'brstuff');
        // $this->loadViewsFrom(__DIR__.'/../resources/views', 'brstuff');
        // $this->loadMigrationsFrom(__DIR__.'/../database/migrations');
        // $this->loadRoutesFrom(__DIR__.'/routes.php');

        // Publishing is only necessary when using the CLI.
        if ($this->app->runningInConsole()) {
            $this->bootForConsole();
        }
    }

    /**
     * Register any package services.
     *
     * @return void
     */
    public function register()
    {
        $this->mergeConfigFrom(__DIR__.'/../config/laravelsqlines.php', 'laravelsqlines');

        // Register the service the package provides.
        $this->app->singleton('laravelsqlines', function ($app) {
            return new LaravelSqlines;
        });
    }

    /**
     * Get the services provided by the provider.
     *
     * @return array
     */
    public function provides()
    {
        return ['laravelsqlines'];
    }
    
    /**
     * Console-specific booting.
     *
     * @return void
     */
    protected function bootForConsole()
    {
        // Publishing the configuration file.
        $this->publishes([
            __DIR__.'/../config/laravelsqlines.php' => config_path('laravelsqlines.php'),
        ], 'laravelsqlines.config');

        // Publishing the views.
        /*$this->publishes([
            __DIR__.'/../resources/views' => base_path('resources/views/vendor/brstuff'),
        ], 'laravelsqlines.views');*/

        // Publishing assets.
        /*$this->publishes([
            __DIR__.'/../resources/assets' => public_path('vendor/brstuff'),
        ], 'laravelsqlines.views');*/

        // Publishing the translation files.
        /*$this->publishes([
            __DIR__.'/../resources/lang' => resource_path('lang/vendor/brstuff'),
        ], 'laravelsqlines.views');*/

        // Registering package commands.
        // $this->commands([]);
        
        $this->publishes([
            __DIR__.'/../resources/sqlines' => storage_path('laravelsqlines'),
        ], 'laravelsqlines.app');
        
    }
}


src/LaravelSqlines.php
<?php

namespace brstuff\LaravelSqlines;

use File;

class LaravelSqlines
{
    private $files = [];
    private $sqlines_storage = "laravelsqlines";
    private $sqlines_app = "sqldata";
    
    private $configOptions = [
        'sd',
        'td',
        't',
        'ddl_tables',
        'indexes',
        'data',
        'batch_max_rows' => 10000,
        'fetch_lob_as_varchar',
        'lob_bind_buffer' => 1000000,
        'char_length_ratio',
        'mariadb_lib',
        'mariadb_set_sql_log_bin' => 0,
        'mariadb_set_unique_checks' => 0,
        'mariadb_max_allowed_packet',
        'mysql_set_character_set' => 'utf8',
        'mysql_set_collation' => 'utf8_bin',
        'mysql_set_foreign_key_checks' => 0,
        'mysql_validation_collate' => 'latin1_bin',
        'bcp_codepage',
        'oci_lib',
        'oracle_nls_lang' => 'AMERICAN_AMERICA.WE8ISO8859P1',
        'pg_client_encoding',
        'sybase_codepage' => 'cp850',
        'sybase_encrypted_password',
        'sybase_asa_char_as_wchar',
        'informix_client_locale',
        'validation_not_equal_max_rows',
        'validation_datetime_fraction',
        'trace',
        'trace_data',
        'trace_diff_data',
    ];
    
    /**
     *  todo:
     *  cmd
     *  queries
     *  etc  
     */
    private $options = [
        'source' => '',
        'target' => '',
        'tables' => [],
        'tables_excluded' => [],
        'sql_selected' => [],
        'command' => '',    // TODO
        'topt' => [],
        'vopt' => '',
        'smap' => [],
        'cmap' => [], 
        'dtmap' => [],
        'tmap' => [],
        'queries' => [],    // TODO
        'twhere' => [],
        'tsel' => [],
        'tsel_all' => [],
    ]; 
    
    private $connections = ['oracle','sql','db2','mariadb','mysql','pg','sybase','asa','informix'];
    
    /**
     * Source
     * 
     * @param string $connection (oracle|sql|db2|mariadb|mysql|pg|sybase|asa|informix)
     * @param string $host Host of database
     * @param string $username Username of database
     * @param string $password Password of database
     * @param string $database Name of database
     * @param string $port Port
     * @throws \Exception
     * @return $this
     */
    public function source($connection, $username, $password = "", $database, $host = "", $port=""){
        $this->options['source'] = $this->connectionString($connection, $username, $password, $database, $host, $port);
        return $this;
    }
    
    /**
     * Target
     *
     * @param string $connection (oracle|sql|db2|mariadb|mysql|pg|sybase|asa|informix)
     * @param string $host Host of database
     * @param string $username Username of database
     * @param string $password Password of database
     * @param string $database Name of database
     * @param string $port Port
     * @throws \Exception
     * @return $this
     */
    public function target($connection, $username, $password = "", $database, $host = "", $port=""){
        $this->options['target'] = $this->connectionString($connection, $username, $password, $database, $host, $port);
        return $this;
    }
    
    /**
     * sourceConnection
     * 
     * @param string $connection (oracle|sql|db2|mariadb|mysql|pg|sybase|asa|informix)
     * @param string $database_config config database name
     * @return $this
     */
    public function sourceConnection($connection, $database_config){
        $cfg = config("database.connections.".$database_config);
        $this->source($connection, $cfg['username'],  $cfg['password'],  $cfg['database'],  $cfg['host'],  $cfg['port']);
        return $this;
    }
    
    /**
     * targetConnection
     *
     * @param string $connection (oracle|sql|db2|mariadb|mysql|pg|sybase|asa|informix)
     * @param string $database_config config database name
     * @return $this
     */
    public function targetConnection($driver, $connection){
        $cfg = config("database.connections.".$connection);
        $this->target($driver, $cfg['username'],  $cfg['password'],  $cfg['database'],  $cfg['host'],  $cfg['port']);
        return $this;
    }
    
    /**
     * Tables to sync
     * 
     * @param string|array $tables Can be single table, array of tables, or splited with commas
     * return $this
     */
    public function tables($tables){
        $tables = $this->textComasOrArray2Array($tables);
        
        if (empty($tables)) { 
            throw new \Exception('Table(s) empty.'); 
        }
         
        $this->options['tables'] = array_merge($this->options['tables'], $tables);
        return $this;
    }
    
    /**
     * Tables to excluded
     *
     * @param string|array $tables Can be single table, array of tables, or splited with commas
     * return $this
     */
    public function tables_excluded($tables){
        $tables = $this->textComasOrArray2Array($tables);
        
        if (empty($tables)) { 
            throw new \Exception('Table(s) empty.'); 
        }
        
        $this->options['tables_excluded'] = array_merge($this->options['tables_excluded'], $tables);
        return $this;
    }
    
    /**
     * CMD
     * 
     * @param string $option Option can be transfer, validate or assess
     * @throws \Exception
     * @return $this
     */
    public function cmd($option){
        $option = strtolower($option);
        
        if (in_array($option, ['transfer', 'validate', 'assess'])) { 
            throw new \Exception('The cmd need to be transfer, validate or assess.'); 
        }
        
        $this->options['command'] = $option;
        return $this;
    }
    
    /**
     * topt
     *
     * return $this
     */
    public function topt($actions){
        $options = ['create','recreate','truncate','pnone'];
        
        $actions = $this->textComasOrArray2Array($actions);
        
        $actions = array_unique(array_intersect($actions, $options));
        
        if (empty($actions)) {
            throw new \Exception('topt empty.');
        }
        
        $this->options['topt'] = $actions;
        return $this;
    }
    
    /**
     * vopt
     *
     * @throws \Exception
     * @return $this
     */
    public function vopt($option){
        $option = strtolower($option);
        
        if (in_array($option, ['rowcount', 'rows'])) {
            throw new \Exception('The vopt need to be rowcount or rows.');
        }
        
        $this->options['vopt'] = $option;
        return $this;
    }
    
    
    
    /**
     * SMAP
     * option specifies the schema name mapping. For example, if you are migrating all tables from schema scott in Oracle, and want to move them to dbo schema in SQL Server, specify -smap = scott:dbo
     * 
     * @throws \Exception
     * @return $this
     */
    public function smap($text){
        $text = $this->textComasOrArray2Array($text);
        
        if (empty($text)) {
            throw new \Exception('smap empty.');
        }
        
        $this->options['smap'] = array_merge($this->options['smap'], $text);
        return $this;
    }
    
    /**
     * cmap
     * option specifies an ASCII file containing a column name and data type mapping. 
     *
     * @throws \Exception
     * @return $this
     */
    public function cmap($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('cmap empty.');
        }
        
        $this->options['cmap'] = array_merge($this->options['cmap'], $text);
        return $this;
    }
    
    /**
     * dtmap
     * option specifies an ASCII file containing a column name and data type mapping.
     *
     * @throws \Exception
     * @return $this
     */
    public function dtmap($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('dtmap empty.');
        }
        
        $this->options['dtmap'] = array_merge($this->options['dtmap'], $text);
        return $this;
    }
    
    
    /**
     * tmap
     * Table name mapping file. Example:
     * SALES.CONTACTS, CRM.SALES_CONTACTS
     *
     * @throws \Exception
     * @return $this
     */
    public function tmap($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('tmap empty.');
        }
        
        $this->options['tmap'] = array_merge($this->options['tmap'], $text);
        return $this;
    }
    
    
    /**
     * queries
     * SQL SELECT queries 
     *
     * @throws \Exception
     * @return $this
     */
    public function queries($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('queries empty.');
        }
        
        $this->options['queries'] = array_merge($this->options['queries'], $text);
        return $this;
    }
    
    
    /**
     * twhere
     * 
     * WHERE conditions for tables. Example :
     * SALES.ORDERS, created_dt >= CURRENT_DATE; 
     *
     * @throws \Exception
     * @return $this
     */
    public function twhere($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('twhere empty.');
        }
        
        $this->options['twhere'] = array_merge($this->options['twhere'], $text);
        return $this;
    }
    
    /**
     * tsel
     * 
     * Select expressions for tables. Example:
     * SALES.CONTACTS, NAME, SUBSTR(CREATED_DT, 1, 10) AS CREATED_DT, 'HQ OFFICE' AS OFFICE;
     *
     * @throws \Exception
     * @return $this
     */
    public function tsel($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('tsel empty.');
        }
        
        $this->options['tsel'] = array_merge($this->options['tsel'], $text);
        return $this;
    }
    
    /**
     * tsel_all
     * 
     * Select expressions for all tables (not applied for tables defined in sqlines_tsel). Example:
     * *, 'NA REGION' AS REGION;   
     *
     * @throws \Exception
     * @return $this
     */
    public function tsel_all($text){
        $text = is_array($text) ? $text : [$text];
        
        if (empty($text)) {
            throw new \Exception('tsel_all empty.');
        }
        
        $this->options['tsel_all'] = array_merge($this->options['tsel_all'], $text);
        return $this;
    }
    
    
    
    
    
    
    public function sync(){
        $export_paths = "";
        $sqlines_path = storage_path($this->sqlines_storage)."/";
        $tmp_prefix = uniqid("tmp_");

        
        $this->files = [
            'app'       => $this->sqlines_storage."/".$this->sqlines_app,
            'cfg'       => $this->sqlines_storage."/".$tmp_prefix."_cfg.tmp",
            'tables'    => $this->sqlines_storage."/".$tmp_prefix."_tables.tmp",
            'tmap'      => $this->sqlines_storage."/".$tmp_prefix."_tmap.tmp",
            'cmap'      => $this->sqlines_storage."/".$tmp_prefix."_cmap.tmp",
            'twhere'    => $this->sqlines_storage."/".$tmp_prefix."_twhere.tmp",
            'cnsmap'    => $this->sqlines_storage."/".$tmp_prefix."_cnsmap.tmp",
            'tsel'      => $this->sqlines_storage."/".$tmp_prefix."_tsel.tmp",
            'tsel_all'  => $this->sqlines_storage."/".$tmp_prefix."_tsel_all.tmp",
        ];
        
        if (empty($this->options['source'])) {
            throw new \Exception('Source is empty.');
        }
        if (empty($this->options['target'])) {
            throw new \Exception('Target is empty.');
        }
        if (empty($this->options['tables'])) {
            throw new \Exception('Tables is empty.');
        }
            
        if(!File::exists(storage_path($this->files['app']))) {
            throw new \Exception('SQlines App not present in: '.storage_path($this->files['app'])."\nTry run: php artisan vendor:publish --tag=laravelsqlines.app");
        }
        
        if(!empty(config('laravelsqlines.export_paths'))){
            $export_paths = sprintf('export LD_LIBRARY_PATH=\$LD_LIBRARY_PATH:%s; ', config('laravelsqlines.export_paths'));
        }
        
        $options = [];
        $options[] = $this->option2File("tmap", "tmapf");
        $options[] = $this->option2File("cmap", "cmapf");
        $options[] = $this->option2File("tables", "tf");
        $options[] = $this->option2File("tsel", "tself");
        $options[] = $this->option2File("cnsmap", "cnsmapf");
        $options[] = $this->option2File("tsel_all", "tsel_allf");
        $options = array_filter($options);
        
        $this->config2File($options);
        $conn = sprintf("%s %s -cfg=%s", $export_paths, storage_path($this->files['app']), storage_path($this->files['cfg']));
        $log = shell_exec($conn);
        $this->deleteTmp();
        return $log;
    }
    
    private function config2File($optionsfiles = []){
        $config = array_filter(config('laravelsqlines.data-options'));
        $config['sd'] = $this->options['source'];
        $config['td'] = $this->options['target'];
        
        $cfg_content = empty($optionsfiles) ?  "" : implode(" \n", $optionsfiles)."\n";
        foreach ($config as $key => $value){
            $cfg_content .= sprintf("-%s=%s \n", $key, $value);
        }
        
        $file = storage_path($this->files['cfg']);
        file_put_contents($file, $cfg_content);
    }
    
    private function option2File($option, $optionf){
        if(!empty($this->options[$option])){
            file_put_contents(storage_path($this->files[$option]), implode("\n", $this->options[$option]));
            return sprintf("-%s=%s", $optionf, storage_path($this->files[$option]));
        }
        return "";
    }
    
    
    private function textComasOrArray2Array($text){
        if(is_array($text)){
            $array = $text;
        }elseif( strpos($text, ",") !== false ) {
            $array = explode(",", $text);
        }else{
            $array = [$text];
        }
        return $array;
    }
    
    private function deleteTmp(){
        $sqlines_path = storage_path($this->sqlines_storage)."/";
        array_map('unlink', glob( $sqlines_path."*.{tmp,txt,sql,log}", GLOB_BRACE));
    }
    
    
    private function connectionString($connection, $username, $password = "", $database, $host = "", $port="")
    {
        if (!in_array($connection, $this->connections)) {
            throw new \Exception('Invalid type of connection.');
        }
        
        switch ($connection){
            case 'oracle':
                $host = empty($port) ? $host : $host.':'.$port;
                $credentials = empty($password) ? $username : $username."/".$password;
                $connection_string = sprintf("%s, %s@%s/%s", $connection, $credentials, $host, $database);
                break;
                
            case 'asa':
            case 'db2':
                $credentials = empty($password) ? $username : $username."/".$password;
                $connection_string = sprintf("%s, %s@%s", $connection, $credentials, $database);
                break;
                
            default:
                $host = empty($port) ? $host : $host.':'.$port;
                $credentials = empty($password) ? $username : $username."/".$password;
                $connection_string = sprintf("%s, %s@%s,%s", $connection, $credentials, $host, $database);
                break;
        }
        
        return $connection_string;
    }
    
    
}